pkg/controller/remotecluster/controller.go

```diff
@@ -0,0 +1,250 @@
+	runtimeutil.Must(networkingv1.AddToScheme(scheme.Scheme))
+
+	klog.V(4).Info("Creating event broadcaster")
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(klog.Infof)
+	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: kubeClient.CoreV1().Events("")})
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: ControllerName})
+
+	uuid, err := utils.GetUUID(kubeClient)
+	if err != nil {
+		panic(err)
+	}
+
+	c := &Controller{
+		rcMgrCache:                NewCache(),
+		UUID:                      uuid,
+		kubeClient:                kubeClient,
+		ramaClient:                ramaClient,
+		remoteClusterLister:       remoteClusterInformer.Lister(),
+		remoteClusterSynced:       remoteClusterInformer.Informer().HasSynced,
+		remoteSubnetLister:        remoteSubnetInformer.Lister(),
+		remoteSubnetSynced:        remoteSubnetInformer.Informer().HasSynced,
+		localClusterSubnetLister:  localClusterSubnetInformer.Lister(),
+		localClusterSubnetSynced:  localClusterSubnetInformer.Informer().HasSynced,
+		remoteVtepLister:          remoteVtepInformer.Lister(),
+		remoteVtepSynced:          remoteSubnetInformer.Informer().HasSynced,
+		localClusterNetworkLister: localClusterNetworkInformer.Lister(),
+		localClusterNetworkSynced: localClusterNetworkInformer.Informer().HasSynced,
+		remoteClusterQueue:        workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), ControllerName),
+		rcMgrQueue:                workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "remoteclustermanager"),
+		recorder:                  recorder,
+	}
+
+	remoteClusterInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: c.filterRemoteCluster,
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addOrDelRemoteCluster,
+			UpdateFunc: c.updateRemoteCluster,
+			DeleteFunc: c.addOrDelRemoteCluster,
+		},
+	})
+
+	return c
+}
+
+func (c *Controller) Run(stopCh <-chan struct{}) error {
+	defer runtimeutil.HandleCrash()
+	defer c.rcMgrQueue.ShutDown()
+	defer c.remoteClusterQueue.ShutDown()
+
+	klog.Infof("Starting %s controller", ControllerName)
+
+	klog.Info("Waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, c.remoteClusterSynced, c.remoteSubnetSynced, c.remoteVtepSynced, c.localClusterSubnetSynced); !ok {
+		return fmt.Errorf("%s failed to wait for caches to sync", ControllerName)
+	}
+
+	// start workers
+	klog.Info("Starting workers")
+	go wait.Until(c.runRemoteClusterWorker, time.Second, stopCh)
+	go wait.Until(c.processRCManagerQueue, time.Second, stopCh)
+	go wait.Until(c.runOverlayNetIDWorker, time.Minute, stopCh)
+	go wait.Until(c.updateRemoteClusterStatus, HealthCheckPeriod, stopCh)
+	<-stopCh
+
+	c.closeRemoteClusterManager()
+
+	klog.Info("Shutting down workers")
+	return nil
+}
+
+func (c *Controller) closeRemoteClusterManager() {
+	// no need to lock
```

Comment:
why no need to lock here? stopCh is just a signal, it can not make sure every working threads have already exited successfully

Reply:
Do we really need to lock for close? This close is used to close manager's informer factory, I think close it is enough. If locked, there maybe some delay because of the lock competion. What's more, even the goroutine exited failed, the main thread will do the clean job.

Reply:
If not block here, the main thread will **force** clean all goroutines, which means more unexpected situations will happen.

Reply:
ok